zerostash-files/src/tree.rs

```diff
@@ -0,0 +1,674 @@
+        self.insert_root().unwrap();
+        let Some(noderef) = self.get_ref(path)? else {
+            return Ok(None);
+        };
+
+        let Some(node) = self.0.get(&noderef) else {
+            return Err(FsError::InvalidFilesystem);
+        };
+
+        Ok(node.as_file())
+    }
+
+    /// Return a file system node
+    pub fn get<'a>(&self, path: &'a str) -> Result<'a, Option<Arc<Node>>> {
+        self.insert_root().unwrap();
+        let Some(noderef) = self.get_ref(path)? else {
+            return Ok(None);
+        };
+
+        let Some(node) = self.0.get(&noderef) else {
+            return Err(FsError::InvalidFilesystem);
+        };
+
+        Ok(Some(node))
+    }
+
+    /// Move the file from the old path to the new path in the tree
+    pub fn move_node<'a>(&self, old_path: &'a str, new_path: &'a str) -> Result<'a, ()> {
+        self.insert_root().unwrap();
+        let (parent_ref, _, node_name) = self.path_to_parent(old_path)?;
+        let noderef = {
+            let mut noderef = None;
+            self.0.update_with(parent_ref, |v| {
+                let current = v.as_ref().clone();
+                let Node::Directory { ref entries } = current else {
+                    unreachable!()
+                };
+                noderef = entries.remove(node_name);
+                current
+            });
+
+            let Some((_, noderef)) = noderef else {
+                return Err(FsError::NoSuchFileOrDirectory);
+            };
+            noderef
+        };
+
+        // add to the new place, overwriting an existing file there
+        let (new_ref, _, new_node_name) = self.path_to_parent(new_path)?;
+        self.0.update_with(new_ref, |v| {
+            let new = v.as_ref().clone();
+            let Node::Directory { ref entries } = new else {
+                unreachable!()
+            };
+            _ = entries.insert(new_node_name.into(), noderef);
+            new
+        });
+
+        Ok(())
+    }
+
+    pub fn root(&self) -> Arc<Node> {
+        self.0.get(&Digest::default()).expect("uninitialized tree")
+    }
+
+    fn remove_file(&self, parent_ref: &Digest, filename: &str) {
+        self.0.update_with(*parent_ref, |v| {
+            let new = v.as_ref().clone();
+            if let Node::Directory { ref entries } = new {
+                entries.remove(filename);
+                return new;
+            }
+            unreachable!()
+        });
+    }
+
+    /// Return the internal reference to the path
+    fn get_ref<'a>(&self, path: &'a str) -> Result<'a, Option<Digest>> {
+        let (_, current, filename) = self.path_to_parent(path)?;
+        let Node::Directory { ref entries } = current.as_ref() else {
+            unreachable!()
+        };
+
+        Ok(entries.read(filename, |_, v| *v))
+    }
+
+    /// Returns if the path is a file
+    pub fn is_file<'a>(&self, path: &'a str) -> Result<'a, bool> {
+        self.insert_root().unwrap();
```

Comment:
I don't think this line should be required to call at the start of every function. Any reason we can't move this into a `impl Default for Tree {}`?

Reply:
I think its a bug with the VersionedMap, although Im not entirely certain. I just now tried to look for the reason why it is not working, and it seems like after reopening, the Default implementation works fine. However, if I now load the full tree into memory, all content gets added to the tree, causing it to contain 2 Default Digests, which then causes issues